charms/reactive/flags.py

```diff
@@ -83,6 +85,8 @@ def set_flag(flag, value=None):
             set_flag(flag_name)
         for flag_name in trigger['clear_flag']:
             clear_flag(flag_name)
+        for callback in trigger['callbacks']:
+            callback()
```

Comment:
I'm slightly nervous about the callback being triggered within the handling of the current handler, rather than being scheduled as the next handler to run?  But, to be fair, I've not fully understood how the callbacks are supposed to interact with handlers in terms of migrated interfaces to op-framework code?

Reply:
The callback isn't intended to be a handler, and should really be limited to a minimal operation akin to setting or clearing a flag.

The specific use-case I had in mind for this is an `is_changed` property which is mapped to the `endpoint.{relation_name}.changed` flag, where clearing the flag would then automatically invoke `inst.is_changed = False` which would presumably update a hash value of relevant data in stored state so that the change is considered "handled." This is [handled automatically by the library using this feature](https://github.com/juju-solutions/ops-reactive-interface/pull/1/files#diff-be35484ccd89c06a5d619f4c4cbfc0aa1c541b7152f79f255b75178c566e12c5R121-R126) so that the interface APIs can just keep all charms.reactive-related code in a single `manage_flags` method (since that is guaranteed to only be invoked when used in a reactive charm) and otherwise assume that changes will be acknowledged via the property. Then, reactive charms can interact with the flag in the natural "reactive" fashion: have a handler triggered by the flag, handle the change, and clear the flag to indicate that it's done, without having to worry about forgetting to clear the property as well.

However, I'm thinking that this may be a bit too much magic behavior and maybe it would be safer in the end to just put a bit more onus on the reactive developer by saying that the `.changed` flag will be set automatically, but it's up to the charm to clear it and if the specific interface API also requires a property to be cleared, you'll have to remember to do that as well.

Reply:
I think, based on this discussion, that the callback should be kept.  If a pattern can be established for migration of interfaces to ops-framework with interoperability back to charms.reactive code, then a little bit of magic is a fine thing :)  So I'm broadly in favour of that and the code.